### PR TITLE
Created Legacy namespaces with copies of ...

### PR DESCRIPTION
**Pull request by marcfehling**
... SolutionTransfer, FEFieldFunction, DataOut, DataOut_DoFData, DataOutFaces, DataOutRotation.

Part of #10333.

Removing the `DoFHandlerType` template argument in these classes is not trivial, as we still need the `spacedim` argument. Since "overloading" template arguments is not possible, we could think about giving these classes new names, but since they are so fundamentally rooted with the `deal.II` library a lot of users wouldn't be happy about (and we would need to change a lot of tutorial programs for this).

So I propose to go a different route: Create a new namespace `DoFHandlerTypeLegacy` (name open for discussion) which features the old interface to which user can fall back, and provide the affected classes with the new `spacedim` argument.

We may split this over two releases: In the next release, we would introduce the legacy namespace and warn users about the upcoming change in the changelog. For the succeeding release, we would actually change the template arguments. What do you think?

### CODE REVIEW

File: include/deal.II/distributed/solution_transfer.h
Comment (peterrum) [inline include/deal.II/distributed/solution_transfer.h:400]: I am a bit confused why `DoFHandlerType` is passed to the non-legacy version. I guess remove `DoFHandlerType`  there will the done follow-up PRs?

**Review by peterrum — APPROVED** [review]
I like the idea. But I wouldn't create a namespace only dedicated to `DoFHandler` but be more general `Legacy`.

**Comment by marcfehling** [comment]
I enhanced the `sed` script in the changelog entry, which `doxygen` does not like.
```
/home/runner/work/dealii/dealii/build/doc/news/changes.h:929: warning: reached end of file while inside a f$ block!
The command that should end the block seems to be missing!
```

I guess it gets confused with the shell variables accessor symbol `$`. Do you know how to circumvent it?

**Comment by marcfehling** [comment]
Escaping the `$` character did the trick. Although the script begin copied directly from the textfile wouldn't work anymore, it does when being copied from the web version.
changelog

**Comment by masterleinad** [comment]
> We may split this over two releases: In the next release, we would introduce the legacy namespace and warn users about the upcoming change in the changelog. For the succeeding release, we would actually change the template arguments. What do you think?

I wouldn't expect any user to read the changelog but, of course, we need a deprecation changelog entry anyway. I don't quite see the point of offering a separate fallback option with a `Legacy` namespace. If users need to change their code anyway, they should just fix the template arguments and not go to another namespace that we intend to remove at some point again.
In my opinion, we should discuss if this will just be a hard incompatibility or if we want to choose new names (with proper deprecations).

File: include/deal.II/distributed/solution_transfer.h
Comment (bangerth) [inline include/deal.II/distributed/solution_transfer.h:387]: This approach doesn't work well for classes that have non-trivial constructors. That's because when you create an object of this type, you want to call the constructors that existed in the old class, but your new class doesn't have these constructors.

The way to do this is a `using` directive ("type alias"):
```
template <int dim,
￼                typename VectorType,
￼                typename DoFHandlerType = DoFHandler<dim>>
￼ using SolutionTransfer DEAL_II_DEPRECATED 
 = public dealii::parallel::distributed::
￼            SolutionTransfer<dim, VectorType, DoFHandlerType>;
```

File: include/deal.II/distributed/solution_transfer.h
Comment (bangerth) [inline include/deal.II/distributed/solution_transfer.h:394]: I'd much rather we have just one top-level `namespace Legacy` in which we replicate the namespaces from which we draw.

File: include/deal.II/numerics/data_out.h
Comment (bangerth) [inline include/deal.II/numerics/data_out.h:528]: Ah, that's another way -- but I think that the type alias is still the better choice.

**Comment by marcfehling** [comment]
> > We may split this over two releases: In the next release, we would introduce the legacy namespace and warn users about the upcoming change in the changelog. For the succeeding release, we would actually change the template arguments. What do you think?
> 
> I wouldn't expect any user to read the changelog but, of course, we need a deprecation changelog entry anyway. I don't quite see the point of offering a separate fallback option with a `Legacy` namespace. If users need to change their code anyway, they should just fix the template arguments and not go to another namespace that we intend to remove at some point again.
> In my opinion, we should discuss if this will just be a hard incompatibility or if we want to choose new names (with proper deprecations).

I see your point. Users upgrading to a new release may expect incompatibilities, and the changes to be done for the new template arguments are not that fundamental. In fact, I expect most users to use the default template arguments which will not require them to change anything.

I am not a fan of introducing new names for these basic classes, so I would be for either the hard incompatibility or the `Legacy` namespace. It may not hurt to have the latter though.

**Comment by bangerth** [comment]
We've done this in the past when we needed to be incompatible: Introduce an alias for the next release (9.3), remove or change the current function in the release after that (9.4), and remove the alias after that (9.5). This way, users can write programs that can work on two subsequent releases. I think that's a reasonable model for those who don't compile deal.II themselves and for whatever reason want to work on different systems that have different deal.II versions installed.

I find the pain involved in this relatively minor, and we haven't had complaints in the past about incompatible changes (unlike, for example, the PETSc people who get these complaints all the time). I can't say that we don't have these complaints because we're generally careful, or because we use a well-defined model for such changes (namely, having ways to be compatible with two subsequent releases). In any case, I find it useful to be able to point to a strategy.

File: doc/news/changes/incompatibilities/20201120Fehling
Comment (bangerth) [inline doc/news/changes/incompatibilities/20201120Fehling:23]: This seems exceedingly technical for relatively little benefit -- it seems like black magic. You can leave it in, but in general people will likely only have a handful of occurrences of these classes in their code base, and it might be easier to just have a list of replacements, like
```
  If for some reason, you need a code that is compatible with deal.II 9.3 and the subsequent release, you
  can make the following substitutions:
      X -> Legacy::X
      ...
```

File: include/deal.II/numerics/data_out.h
Comment (bangerth) [inline include/deal.II/numerics/data_out.h:526]: On second though, you mark these names as deprecated. But that means that they will disappear in the following release. But that specifically isn't what we want, do we? Shouldn't these names be currently non-deprecated, be marked as deprecated *after* the upcoming release, and removed after the one after that?

File: source/numerics/fe_field_function.cc
Comment (bangerth) [inline source/numerics/fe_field_function.cc:39]: :-O

File: source/numerics/fe_field_function.inst.in
Comment (bangerth) [inline source/numerics/fe_field_function.inst.in:22]: Better, yes. This could also have been done in a separate PR, but let's see how long it takes to get this merged.

File: include/deal.II/numerics/data_out.h
Comment (marcfehling) [reply include/deal.II/numerics/data_out.h:526]: As these will get removed at some point, I thought that keeping the deprecation flags would be a good idea. However, it could happen that they'll get accidentally removed for the next release because of this flag. So yes, let's remove them for now.

I'll open a separate draft PR which deals with the deprecation of these aliases, which I will unblock after the next release.

File: source/numerics/fe_field_function.inst.in
Comment (marcfehling) [reply source/numerics/fe_field_function.inst.in:22]: This is a historic remnant of this PR :)

It feels out of place now. I can move it to a separate PR.

Comment (bangerth) [reply source/numerics/fe_field_function.inst.in:22]: I tend to break these things out of my PRs because they're uncontroversial and easy to merge. But let's not make our lives unnecessarily complicated and just leave it here.

File: include/deal.II/numerics/data_out.h
Comment (bangerth) [reply include/deal.II/numerics/data_out.h:526]: Sounds good.

**Comment by bangerth** [comment]
Let us know when you want me to take another look.

@masterleinad -- you were unconvinced. What's your opinion following the discussion above?

**Comment by masterleinad** [comment]
> We've done this in the past when we needed to be incompatible: Introduce an alias for the next release (9.3), remove or change the current function in the release after that (9.4), and remove the alias after that (9.5). This way, users can write programs that can work on two subsequent releases. I think that's a reasonable model for those who don't compile deal.II themselves and for whatever reason want to work on different systems that have different deal.II versions installed.

Fair enough. That sounds reasonable.

**Comment by marcfehling** [comment]
Applied your suggestions. Ready for final review, after which I will squash all commits.

**Review by bangerth — APPROVED** [review]
Looks good.

In the future, can I ask you to just always squash/fixup commits whenever you update a patch and you think it might be ready? This way we can hit "merge" right after approval, without having to do one more go-around.

**Pull request merged**
